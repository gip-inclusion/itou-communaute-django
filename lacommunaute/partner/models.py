from django.conf import settings
from django.db import models
from django.urls import reverse
from django.utils.encoding import force_str
from django.utils.text import slugify
from django.utils.translation import gettext_lazy as _
from machina.models.abstract_models import DatedModel
from machina.models.fields import MarkupTextField
from storages.backends.s3boto3 import S3Boto3Storage

from lacommunaute.utils.validators import validate_image_size


class Partner(DatedModel):
    name = models.CharField(max_length=100, verbose_name="Nom", unique=True)
    slug = models.SlugField(max_length=255, verbose_name="Slug")
    short_description = models.CharField(
        max_length=400, blank=True, null=True, verbose_name="Description courte (SEO)"
    )
    description = MarkupTextField(verbose_name="Description", null=True, blank=True)
    logo = models.ImageField(
        storage=S3Boto3Storage(bucket_name=settings.AWS_STORAGE_BUCKET_NAME, file_overwrite=False),
        validators=[validate_image_size],
        upload_to="logos/",
        help_text="1200x600 recommand√©",
    )
    url = models.URLField(verbose_name="Lien vers le site du partenaire", null=True, blank=True)

    objects = models.Manager()

    class Meta:
        verbose_name = _("Partner")
        verbose_name_plural = _("Partners")
        ordering = ["-updated"]

    def __str__(self):
        return f"{self.name}"

    def save(self, *args, **kwargs):
        self.slug = slugify(force_str(self.name), allow_unicode=True)
        super().save(*args, **kwargs)

    def get_absolute_url(self):
        return reverse("partner:detail", kwargs={"slug": self.slug, "pk": self.pk})
