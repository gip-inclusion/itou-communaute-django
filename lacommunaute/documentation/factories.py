import factory
from faker import Faker

from lacommunaute.documentation.models import Category, Document, DocumentRating


faker = Faker()


class AbstractDocumentationFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    description = factory.Faker("sentence", nb_words=100)
    short_description = factory.Faker("sentence", nb_words=10)
    image = factory.django.ImageField(filename="banner.jpg")


class CategoryFactory(AbstractDocumentationFactory):
    class Meta:
        model = Category

    class Params:
        for_snapshot = factory.Trait(
            name="Test Category", description="Test description", short_description="Test short description"
        )


class DocumentFactory(AbstractDocumentationFactory):
    category = factory.SubFactory(CategoryFactory)

    class Meta:
        model = Document

    class Params:
        for_snapshot = factory.Trait(
            name="Test Document",
            description="Test description",
            short_description="Test short description",
            category=factory.SubFactory(CategoryFactory, for_snapshot=True),
        )

    @factory.post_generation
    def with_tags(self, create, extracted, **kwargs):
        if not create or not extracted:
            return

        if isinstance(extracted, list):
            for tag in extracted:
                self.tags.add(tag)


class DocumentRatingFactory(factory.django.DjangoModelFactory):
    document = factory.SubFactory(DocumentFactory)
    rating = factory.Faker("random_int", min=1, max=5)

    class Meta:
        model = DocumentRating
        skip_postgeneration_save = True

    @factory.post_generation
    def set_created(self, create, extracted, **kwargs):
        if extracted:
            self.created = extracted
            self.save()
