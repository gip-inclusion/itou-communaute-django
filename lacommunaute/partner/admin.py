from django.contrib import admin

from lacommunaute.partner.models import Partner


@admin.register(Partner)
class PartnerAdmin(admin.ModelAdmin):
    list_display = ("name", "url", "created")
    search_fields = ("name",)
    list_filter = ("created",)
