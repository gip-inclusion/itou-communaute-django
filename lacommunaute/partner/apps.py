from django.apps import AppConfig


class PartenerAppConfig(AppConfig):
    label = "partner"
    name = "lacommunaute.partner"
    verbose_name = "Partner"
    verbose_name_plural = "Partners"
