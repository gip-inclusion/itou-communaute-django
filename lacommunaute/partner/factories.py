import factory
from faker import Faker

from lacommunaute.partner.models import Partner


faker = Faker()


class PartnerFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = Partner

    name = factory.LazyAttribute(lambda x: faker.company())
    short_description = factory.LazyAttribute(lambda x: faker.text(max_nb_chars=200))
    description = factory.LazyAttribute(
        lambda x: "\n".join([f"### part {i} \n{faker.paragraph(nb_sentences=20)}" for i in range(2)])
    )
    url = factory.LazyAttribute(lambda x: faker.url())

    class Params:
        with_image = factory.Trait(image=factory.django.ImageField(filename="test.jpg"))
        for_snapshot = factory.Trait(
            name="Best Partner Ever",
            description="### h3 long MD description \n lorem ipsum dolor sit amet, consectetur adipiscing elit.",
            short_description="short description for SEO",
            url="https://www.best-partner-ever.com",
        )
