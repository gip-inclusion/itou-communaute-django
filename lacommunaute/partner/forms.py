from django import forms
from django.conf import settings

from lacommunaute.partner.models import Partner


class PartnerForm(forms.ModelForm):
    logo = forms.ImageField(
        required=False,
        label="Logo, format 200 x 200 pixels recommand√©",
        widget=forms.FileInput(attrs={"accept": settings.SUPPORTED_IMAGE_FILE_TYPES.keys()}),
    )

    class Meta:
        model = Partner
        fields = ("name", "short_description", "description", "logo", "url")
