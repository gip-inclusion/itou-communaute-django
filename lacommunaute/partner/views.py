from django.views.generic import DetailView, ListView

from lacommunaute.forum.models import Forum
from lacommunaute.partner.models import Partner
from lacommunaute.utils.perms import forum_visibility_content_tree_from_forums


class PartnerListView(ListView):
    model = Partner
    template_name = "partner/list.html"
    context_object_name = "partners"
    paginate_by = 8 * 3


class PartnerDetailView(DetailView):
    model = Partner
    template_name = "partner/detail.html"
    context_object_name = "partner"

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        context["sub_forums"] = forum_visibility_content_tree_from_forums(
            self.request, Forum.objects.filter(partner=self.object)
        )
        return context
