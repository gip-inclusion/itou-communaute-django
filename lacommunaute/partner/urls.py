from django.urls import path

from lacommunaute.partner.views import PartnerDetailView, PartnerListView


app_name = "partner"


urlpatterns = [
    path("", PartnerListView.as_view(), name="list"),
    path("<str:slug>-<int:pk>/", PartnerDetailView.as_view(), name="detail"),
]
