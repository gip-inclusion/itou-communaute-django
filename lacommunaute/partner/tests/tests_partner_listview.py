import pytest
from django.urls import reverse
from factory import Iterator

from lacommunaute.partner.factories import PartnerFactory
from lacommunaute.users.factories import UserFactory
from lacommunaute.utils.testing import parse_response_to_soup


@pytest.fixture(name="url")
def url_fixture():
    return reverse("partner:list")


def test_listview(client, db, snapshot, url):
    partner = PartnerFactory(for_snapshot=True, with_logo=True)
    response = client.get(url)
    assert response.status_code == 200
    assert str(
        parse_response_to_soup(response, selector="#partner-list", replace_img_src=True, replace_in_href=[partner])
    ) == snapshot(name="partner_listview")


def test_pagination(client, db, snapshot, url):
    num_of_partners = 8 * 3 + 1
    PartnerFactory.create_batch(num_of_partners, name=Iterator([f"Partner {i}" for i in range(num_of_partners)]))
    response = client.get(url)
    assert response.status_code == 200
    assert str(parse_response_to_soup(response, selector="ul.pagination")) == snapshot(name="partner_pagination")


@pytest.mark.parametrize(
    "user,link_is_visible",
    [(None, False), (lambda: UserFactory(), False), (lambda: UserFactory(is_staff=True), True)],
)
def test_link_to_createview(client, db, url, user, link_is_visible):
    if user:
        client.force_login(user())
    response = client.get(url)
    assert response.status_code == 200
    assert bool(reverse("partner:create") in response.content.decode()) == link_is_visible
