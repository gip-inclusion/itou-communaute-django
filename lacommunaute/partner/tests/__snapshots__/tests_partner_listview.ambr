# serializer version: 1
# name: test_listview[partner_listview]
  '''
  <section class="s-section" id="partner-list">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12">
                      <div class="row row-cols-1 row-cols-sm-2 row-cols-lg-3 row-cols-xl-4">
                          
                              <div class="col mb-3 mb-md-5">
                                  <div class="card c-card has-one-link-inside h-100">
                                      
                                          <img alt="Best Partner Ever" class="card-img-top" loading="lazy" src="[img src]"/>
                                      
                                      <div class="card-body pb-0">
                                          <p class="h3 lh-base">Best Partner Ever</p>
                                          <div class="card-text">short description for SEO</div>
                                      </div>
                                      <div class="card-footer text-end">
                                          <a class="btn btn-sm btn-ico btn-link stretched-link matomo-event" data-matomo-action="view" data-matomo-category="engagement" data-matomo-option="partner" href="/partenaires/best-partner-ever-[PK of Partner]/">
                                              <i aria-hidden="true" class="ri-arrow-right-line ri-lg"></i>
                                          </a>
                                      </div>
                                  </div>
                              </div>
                          
                      </div>
                      
                  </div>
              </div>
          </div>
      </section>
  '''
# ---
# name: test_pagination[partner_pagination]
  '''
  <ul class="m-0 pagination pagination-sm justify-content-center mt-5">
          <li class="page-item disabled">
              <a aria-label="Précédent" class="page-link" href=""><i class="ri-arrow-left-s-line"></i></a>
          </li>
          
              
                  <li class="page-item active">
                      <a class="page-link" href="/partenaires/?page=1">1</a>
                  </li>
                  
              
          
              
                  
                  <li class="page-item">
                      <a class="page-link" href="/partenaires/?page=2">2</a>
                  </li>
              
          
          <li class="page-item">
              <a aria-label="Suivant" class="page-link" href="/partenaires/?page=2"><i class="ri-arrow-right-s-line"></i></a>
          </li>
      </ul>
  '''
# ---
